### ISSUE TITLE
Prefer `matchAll`

### ISSUE THREAD

**Issue by brettz9** [sindresorhus/eslint-plugin-unicorn#574]
Not offering a PR here, but just thought such a rule could possibly fit your project...

Namely, preferring `String.prototype.matchAll` to `while` loop `exec` assignments.

As per https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/matchAll#Examples , one can replace this:

```js
let match;
while ((match = regexp.exec(str)) !== null) {
}
```

..with the likes of:

```js
for (const match of str.matchAll(regexp)) {
}
```

I find it quite a bit more elegant and semantic, and it allows avoiding defining a variable in the parent block.

If not, no worries, just thought I'd see if it might pique the interest of others.

**Comment by sindresorhus** [comment]
I like it, but the rule needs to be off by default for now as `.matchAll` requires Node.js 12.

**Comment by fisker** [comment]
I think difficult part of this is how to detect this.

Maybe check any assignment of `.exec()` call in `while` loop?

**Comment by sindresorhus** [comment]
This is now accepted. PR welcome :)

**Comment by aggmoulik** [comment]
Can anyone please provide some test cases for the same ?

**Comment by brettz9** [comment]
Sorry for the delayed reply. Re: test cases, I can mostly just think of the failing case in the OP, as well as this being left as is (passing):

```js
let match;
while ((match = regexp.exec(str)) !== null) {
  // ...
}
return match; // If `match` is used outside of the block, do not report/fix
```
